storage/core/src/main/java/com/hortonworks/streamline/storage/impl/jdbc/JdbcStorageManager.java

```diff
@@ -61,14 +63,17 @@ public JdbcStorageManager(QueryExecutor queryExecutor) {
     @Override
     public void add(Storable storable) throws AlreadyExistsException {
         log.debug("Adding storable [{}]", storable);
-        final Storable existing = get(storable.getStorableKey());
-
-        if(existing == null) {
-            addOrUpdate(storable);
-        } else if (!existing.equals(storable)) {
-            throw new AlreadyExistsException("Another instance with same id = " + storable.getPrimaryKey()
-                    + " exists with different value in namespace " + storable.getNameSpace()
-                    + ". Consider using addOrUpdate method if you always want to overwrite.");
+
+        try {
+            queryExecutor.insert(storable);
+        } catch (StorageException ex) {
+            if (ex.getCause() instanceof SQLIntegrityConstraintViolationException) {
+                throw new AlreadyExistsException("Another instance with same id = " + storable.getPrimaryKey()
```

Comment:
Javadoc says that 
> The subclass of SQLException thrown when the SQLState class value is '23', or under vendor-specified conditions. This indicates that an integrity constraint (foreign key, primary key or unique key) has been violated.
Please consult your driver vendor documentation for the vendor-specified conditions for which this Exception may be thrown.

So, this includes foreign key, primary key or unique key if the driver supports to throw this Exception.

This is kind of best effort to throw `AlreadyExistsException` by catching `SQLIntegrityConstraintViolationException`. We should not check for `get` and then execute `addOrUpdate` as `addOrUpdate` updates on duplicate keys and we do not want to update on duplicate keys for `add` method. It also avoids two calls to DB for inserting data. If the vendor throws some other Exception then that error is thrown back to the caller.